### DISCUSSION TITLE
How to get/mark relatively without using 'focus'

### DISCUSSION THREAD

**Discussion by Myrdden** [i3/i3#4731] [Q&A] [answered]
Right, so I weirdly find the `up|down|left|right` motions to be rather confusing and unintuitive so I've set up my config to operate on the tree itself with `parent|child|next sibling|prev sibling`. This works great for me with navigation, but moving windows/tiles/whatever has been a bit difficult. There's no built-in way of doing essentially `swap with [next|prev] sibling`, so I just have a script:
```
i3 mark move
i3 focus [next|prev] sibling
i3 swap with mark move
i3 [con_mark = 'move'] focus
i3 unmark move
```
Which works, except that the focus and thus the display is jittering around making it happen which is rather obnoxious.

So, how do I either set the mark directly on the next/prev sibling so I can swap them without the focus bouncing around, or get the id/con_id of the next/prev sibling without using `focus` to do the same?

**Comment by Airblader** [comment, marked as answer]
> So, how do I either set the mark directly on the next/prev sibling so I can swap them without the focus bouncing around, or get the id/con_id of the next/prev sibling without using focus to do the same?

You can use the IPC for that. In the tree response you can find the focused (leaf) container, take its sibling and get the ID from there. I reckon this would even be possible with `i3-msg -t get_tree | jq …`, but my jq-fu isn't good enough for that.

**Reply by Myrdden** [reply]
Fair enough. I was hoping to avoid parsing JSON if possible... jq can be rather nightmarish, and I can't imagine building and then traversing a tree every time a command is input would be terribly efficient.

But, in any case, recurring down the first element in `focus` will eventually take you to the focused node. However that doesn't mean that the root `focus` array is an actual path to the focused node... Out of curiosity, what exactly *does* that array represent? What is "focus order"?

Also if I may add a suggestion, something akin to `mark [up|down|left|right|parent|next (sibling)|prev (sibling)]` and `swap with [next|prev] (sibling)` would occur to me (someone who has zero knowledge of your codebase whatsoever) to potentially be a trivial addition, considering the "motions" probably exist already and it's simply a change of "actions"? Or maybe not.

**Reply by Airblader** [reply]
> jq can be rather nightmarish

You don't _need_ to use jq. If you hop into any higher-level language you can work more "nicely" with JSON.

> I can't imagine building and then traversing a tree every time a command is input would be terribly efficient

The CPU cycles your web browser spent on writing the two posts here so far probably outweigh the JSON parsing for quite a while. :upside_down_face: This command isn't run a hundred times a second, but only when you move focus. Not to mention that even if you use the i3 built-in commands, since i3 sends all of these IPC events, it also performs JSON serialization on all of these actions. If you weren't bothered about that so far, parsing JSON on your end won't make the difference.

> Out of curiosity, what exactly does that array represent? What is "focus order"?

It is a list of the children of that container in the order of last focus, i.e. whenever a child is focused, it moves to the top of the list.

> Also if I may add a suggestion […] would occur to me […] to potentially be a trivial addition

Such things are ususally easy to implement, yes, but they make commands significantly more complex as there's many combinations that one might want to use here. If you want to open a feature request, go ahead, but I wouldn't want to accept that unless it generates significant interest by users precisely because it contributes quite a bit of complexity and can easily be implemented outside of i3.

**Reply by Myrdden** [reply]
All good points :D Seems then I just need to be a bit more creative and also remember that other languages exist in the world. Cheers!
